cloudsync/providers/file.py

```diff
@@ -0,0 +1,504 @@
+    blocksize = 4096
+    d = blake2b(digest_size=32)
+    block = dat.read(blocksize)
+    while block:
+        d.update(block)
+        block = dat.read(blocksize)
+    return d.digest()
+
+
+def detect_case_sensitive(tmpdir=None):
+    """Returns true if temp direcory specified is case sensitive, or gettempdir() if unspecified."""
+    if not tmpdir:
+        tmpdir = tempfile.gettempdir()
+    f1 = os.path.join(tmpdir, "tmp." + os.urandom(16).hex())
+    with open(f1, "w") as f:
+        f.write("x")
+    try:
+        f2 = f1.upper()
+        if os.path.exists(f2):
+            return False
+    finally:
+        os.unlink(f1)
+    return True
+
+
+def casedpath(path):
+    """Fixes the case of a file to the name on disk.
+
+    Works only for a path that exists
+    """
+    if is_osx():
+        url = NSURL.fileURLWithPath_(path)  # will be None if path doesn't exist
+        if not url:
+            return path
+        return url.fileReferenceURL().path()
+
+    r = glob.glob(re.sub(r'([^:/\\])(?=[/\\]|$)', r'[\1]', path))
+    return r[0] if r else path
+
+
+def canonicalize_fpath(case_sensitive: bool, full_path: str) -> str:
+    """Fixes the case of a path - parent folder must exist."""
+    if not case_sensitive:
+        # not needed for case sensitive installations
+        return full_path
+
+    if not os.path.exists(full_path):
+        # casedpath function doesn't work if the path doesn't exist
+        fdir, fname = os.path.split(full_path)
+        cp = casedpath(fdir)
+        fp: typing.Optional[str]
+        if cp:
+            fp = os.path.join(cp, fname)
+        else:
+            log.error("unexpected call to canonicalized with missing parent folder", stack_info=True)
+            fp = None
+    else:
+        fp = casedpath(full_path)  # canonicalizes path to an existing file
+
+    if not fp:
+        log.debug("canonicalize doesn't yet support missing parent folders %s", full_path)
+        return full_path
+
+    return fp
+
+
+@dataclass
+class CacheEnt:
+    mtime = 0.0
+    qhash = b''
+    fhash = b''
+
+
+class Watcher(watchdog_events.FileSystemEventHandler):
+    def __init__(self, parent):
+        self.parent = parent
+
+    def on_any_event(self, event):
+        self.parent._on_any_event(event)                # pylint: disable=protected-access
+
+
+class FileSystemProvider(Provider):                     # pylint: disable=too-many-instance-attributes, too-many-public-methods
+    """
+    FileSystemProvider is a provider that uses the filesystem, and full paths as the storage location.
+    """
+    default_sleep = 0.01
+    name = "filesystem"
+    oid_is_path = True
+    case_sensitive = detect_case_sensitive()
+    _max_queue = 10000
+    _test_event_timeout = 1
+    _test_event_sleep = 0.001
+    _test_creds: typing.Dict[str, str] = {}
+    _test_namespace = os.path.join(tempfile.gettempdir(), os.urandom(16).hex())
+
+    def __init__(self, namespace="/"):
```

Comment:
How does this work with windows, wrt `c:` vs `/` as the namespace? I know we have `win_paths=False` as the default in the base class. Is it up to the client to check and set these correctly?

Reply:
fs provider should probably detect it and the case_sens flag at the time the namespace is set

Reply:
added correct detection